mlrun/launcher/base.py

```diff
@@ -48,3 +55,179 @@ def _enrich_runtime(runtime):
+        runspec.spec.scrape_metrics = scrape_metrics
+        runspec.spec.input_path = (
+            workdir or runspec.spec.input_path or runtime.spec.workdir
+        )
+        if runtime.spec.allow_empty_resources:
+            runspec.spec.allow_empty_resources = runtime.spec.allow_empty_resources
+
+        spec = runspec.spec
+        if spec.secret_sources:
+            runtime._secrets = mlrun.secrets.SecretsStore.from_list(spec.secret_sources)
+
+        # update run metadata (uid, labels) and store in DB
+        meta = runspec.metadata
+        meta.uid = meta.uid or uuid.uuid4().hex
+
+        runspec.spec.output_path = out_path or artifact_path or runspec.spec.output_path
+
+        if not runspec.spec.output_path:
+            if runspec.metadata.project:
+                if (
+                    mlrun.pipeline_context.project
+                    and runspec.metadata.project
+                    == mlrun.pipeline_context.project.metadata.name
+                ):
+                    runspec.spec.output_path = (
+                        mlrun.pipeline_context.project.spec.artifact_path
+                        or mlrun.pipeline_context.workflow_artifact_path
+                    )
+
+                if not runspec.spec.output_path and runtime._get_db():
+                    try:
+                        # not passing or loading the DB before the enrichment on purpose, because we want to enrich the
+                        # spec first as get_db() depends on it
+                        project = runtime._get_db().get_project(
+                            runspec.metadata.project
+                        )
+                        # this is mainly for tests, so we won't need to mock get_project for so many tests
+                        # in normal use cases if no project is found we will get an error
+                        if project:
+                            runspec.spec.output_path = project.spec.artifact_path
+                    except mlrun.errors.MLRunNotFoundError:
+                        logger.warning(
+                            f"project {project_name} is not saved in DB yet, "
+                            f"enriching output path with default artifact path: {mlrun.mlconf.artifact_path}"
+                        )
+
+            if not runspec.spec.output_path:
+                runspec.spec.output_path = mlrun.mlconf.artifact_path
+
+        if runspec.spec.output_path:
+            runspec.spec.output_path = runspec.spec.output_path.replace(
+                "{{run.uid}}", meta.uid
+            )
+            runspec.spec.output_path = mlrun.utils.helpers.fill_artifact_path_template(
+                runspec.spec.output_path, runspec.metadata.project
+            )
+
+        runspec.spec.notifications = notifications or runspec.spec.notifications or []
+        return runspec
+
+    @staticmethod
+    def _are_validate_notifications(runobj) -> bool:
```

Comment:
```suggestion
    def _are_valid_notifications(run_object: RunObject) -> bool:
```

Reply:
maybe valid is not the right word here because the notifications may be valid but we don't support them in iterations.
would go with something that says "should or should not notify"